### PR TITLE
Add extension support and first extension, unwrap_newtypes

### LINKED ISSUE #69: Consider adding top-level attributes
**Syntax:** `#![attribute_name]` (like in Rust)

After reading https://ozkriff.github.io/2017-12-01--devlog.html where @ozkriff said that his configs (https://raw.githubusercontent.com/ozkriff/zemeroth_assets/master/objects.ron) aren't looking very nice because of how RON handles newtypes, I thought about adding attributes which allow to configure RON for your needs. Not only would this allow for toggles like `#![unwrap_newtypes]`, but it could also be used to deal with new features. That way, we can just enforce that you need `#![feature_name]` to enable a RON feature, such that a parser without support for that feature can return a simple and understandable error.

### PR DESCRIPTION
**Pull request by torkleyy**
## TODO

* [x] More tests, especially regarding error handling
* [x] discuss TODOs / design
* [x] ~~docs for extensions~~ (postponed)
* [x] add syntax to Sublime plugin

## Fixes

Fixes #69

### CODE REVIEW

File: src/de/error.rs [hunk 1]
```diff
@@ -78,6 +82,8 @@ impl StdError for Error {
                 ParseError::Eof => "Unexpected end of file",
                 ParseError::ExpectedArray => "Expected array",
                 ParseError::ExpectedArrayEnd => "Expected end of array",
+                ParseError::ExpectedAttribute => "Expected an enable attribute",
+                ParseError::ExpectedAttributeEnd => "Expected a closing `]` after the attribute",
```
Comment (torkleyy) [inline src/de/error.rs]: Ah, this could also fail if `)` is missing. The error message should reflect that.

**Comment by torkleyy** [comment]
I don't see which contains method you want me to use? The extensions are
stored in a usize.

**Comment by torkleyy** [comment]
bors try

File: highlight/RON.sublime-syntax [hunk 2]
```diff
@@ -6,6 +6,10 @@ file_extensions:
 scope: source.ron
 contexts:
   main:
+    # extensions
```
Comment (torkleyy) [inline highlight/RON.sublime-syntax]: This could be done better, but I think it's sufficient. Could you try it out with your editor?

**Comment by torkleyy** [comment]
r? @kvark 

Please also check the syntax file, I didn't check it!

File: highlight/RON.sublime-syntax (hunk 2 above)
Comment (kvark) [reply highlight/RON.sublime-syntax]: I tested it. Seems to work fine.

File: src/de/tests.rs [hunk 3]
```diff
@@ -169,3 +169,43 @@ fn forgot_apostrophes() {
...
+    let de: Result<String> = from_str("#![enable(invalid)] \"Hello\"");
+
+    assert_eq!(de, err(ParseError::NoSuchExtension("invalid".to_string()), 1, 18));
+}
+
+#[test]
+fn multiple_attributes() {
+    #[derive(Debug, Deserialize, PartialEq)]
+    struct New(String);
+    let de: Result<New> = from_str("#![enable(unwrap_newtypes)] #![enable(unwrap_newtypes)] \"Hello\"");
```
Comment (kvark) [inline src/de/tests.rs]: this is not supported by the syntax highlighter, I think
Can we require the features to always come as a list in a single pragma? e.g. `#![enable(feature1, feature2)]`

File: src/parse.rs [hunk 4]
```diff
@@ -135,6 +149,21 @@ impl<'a> Bytes<'a> {
         }
     }
 
+    fn consume_all(&mut self, all: &[&str]) -> bool {
+        all
+            .iter()
+            .map(|elem| {
+                if self.consume(elem) {
+                    self.skip_ws();
+
```
Comment (kvark) [inline src/parse.rs]: weird newline

File: src/de/tests.rs (hunk 3 above)
Comment (torkleyy) [reply src/de/tests.rs]: I think it is supported? It should match against the same pattern twice.

File: src/parse.rs (hunk 4 above)
Comment (torkleyy) [reply src/parse.rs]: Personal style, I always make a newline before the return value.

**Comment by kvark** [comment]
:shipit: 
bors r+

**Comment by torkleyy** [comment]
Just noticed two nits and fixed them.

bors r=kvark

**Comment by Boscop** [comment]
@torkleyy The syntax highlighting in Sublime is messed up as soon as I add attributes:

image

**Comment by kvark** [comment]
Could you file that on https://github.com/ron-rs/sublime-ron please?

**Comment by Boscop** [comment]
Done: https://github.com/ron-rs/sublime-ron/issues/6

**Pull request merged**